.github/workflows/test.yml

```diff
@@ -1,40 +1,74 @@
 name: test
-on: [ push, pull_request ]
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+  workflow_dispatch:
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
 
 env:
   UNDERCOVER_CONFIG: "(\"elsa.el\" \"elsa-*.el\" (:report-format lcov) (:send-report nil))"
 
 jobs:
   test:
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        emacs_version:
-          - 27.1
+        os:
+          - ubuntu-latest
+          - macos-latest
+          - windows-latest
+        emacs-version:
           - 27.2
-          - 28.1
           - 28.2
           - snapshot
```

Comment:
I've removed `27.1` and `28.1` since checks for lower minor versions are often unnecessary.

Reply:
fair point and we will save some unnecessary compute